gui_server_ecm_sync.md

```diff
@@ -0,0 +1,56 @@
+# Synchronization Between the GUI and Server ECM
+
+## Goal
+When creating, deleting, or modifying entities/components in the GUI _while paused_,
+GUI changes should be reflected in the server once simulation is resumed, or stepped while paused.
+
+## Approach
+If a user performs actions that modify/create/remove entities while simulation is paused, the ECM on the GUI side is updated directly to reflect these actions.
+The server's ECM does not reflect these user actions yet, but will be notified of the changes that took place once simulation is resumed or stepped while paused.
+When the server ECM is notified of these changes, it will be updated to have any changes made to the GUI ECM (unless the GUI and server are run in the same process, which will be discussed below).
+Once the server ECM is updated, simulation will be carried out as normal.
+
+The image below depicts how changes to the GUI's ECM are propogated to the server's ECM.
+The GUI and server interaction can be done through an event (green arrows), or through a service (red arrow).
+If GUI and server interaction is done through a service, then there's no way to propogate GUI ECM changes to the server.
```

Comment:
This isn't true anymore, right? We're propagating the ECM changes through transport when we press play.

Reply:
We are propagating the ECM changes on play/step only if the new event approach is used, because the event connects to `ign-gazebo`'s `GuiRunner` to receive information from the GUI ECM: https://github.com/ignitionrobotics/ign-gazebo/blob/25df2996cf68649fc0b7cdb898d90f8953d7f005/src/gui/GuiRunner.cc#L153-L154

The interaction via service can be thought of as "legacy" behavior, since this is what was already being done before GUI ECM propagation functionality was added. The GUI/Server interaction via service hasn't changed, so the service still uses the same message type ([`msgs::WorldControl`](https://github.com/ignitionrobotics/ign-msgs/blob/ign-msgs8/proto/ignition/msgs/world_control.proto)), which does not contain `ECM` information. So, in this scenario, there's no way for the GUI to share ECM information with the server: https://github.com/ignitionrobotics/ign-gazebo/blob/25df2996cf68649fc0b7cdb898d90f8953d7f005/src/SimulationRunner.cc#L1140-L1157

So, all of that being said, this statement is true, but perhaps it can be reworded to something like this for clarity:
```
If GUI and server interaction is done through a service, then there's no way to propagate GUI ECM changes to the server.
GUI ECM changes are only propagated to the server if the GUI and server interaction is done through an event.
```

Let me know if that helps at all, or if you have any additional questions.

Reply:
Following up on my previous comment, I reworded this in https://github.com/ignitionrobotics/design/pull/8/commits/dce6cae9cba7926636d5252dc53d37e636d89e25 to (hopefully) remove confusion. Let me know what you think!